### ISSUE TITLE
Project Level 1 Server class

### ISSUE THREAD

**Issue by karintang** [nus-cs2030/2021-s2#112]
# Summary
My Server class takes in an identifier as the constructor. The identifier is private final. However, the status of my server is not private final. Is it allowed? I have 2 statuses in Server, serving and waiting. They are private in my Server class. Will I violate the immutability principle?

**Comment by Onelayer** [comment]
To not violate the immutability principle, you can return a new Server for your two methods serving and waiting. So instead of your Server object being mutable, you return a new Server to ensure immutability while still being able to change the values of the Server object to meet your needs.

**Comment by hewtungyuen** [comment]
From my experience, we are not allowed to have any attributes that are not private final, as it will fail the CodeCrunch test. 

For my implementation, I created an array of servers in my Main class. I also created Event classes (e.g. ArriveEvent, ServeEvent), which has the method nextEvent that takes in a array of servers, modifies the server in the array and returns the next event. By calling .nextEvent(Server[] servers) in the main class, I am hence able to modify the array of servers that is in my Main class.

**Comment by gableejh** [comment]
I guess you can always create a new Server with the new attributes and return the new Server as shared by hewtungyuen, once CodeCrunch detects a variable that is not private/protected final, it'll just be graded at F immediately.

**Comment by ashleyleerx** [comment]
As what the others said you can create a new Server with the updated attributes.  Alternatively, for abstraction purposes, you can include a private constructor along with a method in the Server to update the Server something like updateWaitingTime. For me in my nextEvent method I took in a server array so I updated the specific server in the specific array using a method in the Server class

**Comment by KevinTan1203** [comment]
Hi Karin, I believe the others above have also mentioned it, but yes, you need to ensure all your attributes are private final. If you want to update your serving and waiting attributes, then you will need to return a new Server, and that could be a method under your server class (e.g addWaiting(Person person) { ... }/ addServing(Person person) {...})

**Comment by Keithlimhs** [comment]
Hi as we have learnt the OOP principles, we should always declare attributes as private final to adhere to immutability and encapsulation. Hence if the status of your server is not private final it will raise an error in codecrunch and you would get an F. If you want to update relevant attributes of the Server class while still adhering to the principles, you should have an additional constructor of Server which updates the relevant attributes by creating the new Server, and you can use this constructor in your methods which requires you to change the immutable attributes

**Comment by coding4soup** [comment]
Hi Karin, hope you solved this already!

But anyways, if you have not, then what you could do to create and update the status of your Server is to essentially return a new Server every time.

A small loophole is to create a private final ArrayList, and put the Serving/Occupied status inside the arraylist. ArrayLists are reference types, so the reference is immutable, but the contents can be changed. Although I have no idea if that's really legal.

**Comment by Xeph-o** [comment]
I recommend using an ArrayList of servers, with the .set() method to input a new server with the updated changes. This would however require you to pass another field around - the index of the server in the array.

**Comment by ivankqw** [comment]
> Hi Karin, hope you solved this already!
> 
> But anyways, if you have not, then what you could do to create and update the status of your Server is to essentially return a new Server every time.
> 
> A small loophole is to create a private final ArrayList, and put the Serving/Occupied status inside the arraylist. ArrayLists are reference types, so the reference is immutable, but the contents can be changed. Although I have no idea if that's really legal.

Yup I believe that this does not violate the immutability principle as mentioned by prof in one of the recitations

**Comment by brennanleez-coder** [comment]
Hi karin,
hope you have solved this by now but to add on to the good suggestions above, the only way is to return a new Server everytime u change the status. If you store the severs in an array, you should then have a helper method to duplicate the array, add the elements of the old array and copy it into the new array, then finally change the server in the new array.

**Comment by saiyara1512** [comment]
> From my experience, we are not allowed to have any attributes that are not private final, as it will fail the CodeCrunch test.
> 
> For my implementation, I created an array of servers in my Main class. I also created Event classes (e.g. ArriveEvent, ServeEvent), which has the method nextEvent that takes in a array of servers, modifies the server in the array and returns the next event. By calling .nextEvent(Server[] servers) in the main class, I am hence able to modify the array of servers that is in my Main class.

Hi! I did this, but I have a cyclic dependency between my Event class and Server class. This you experience something similar, and if so how did you fix it? Thanks!

**Comment by hewtungyuen** [comment]
> > From my experience, we are not allowed to have any attributes that are not private final, as it will fail the CodeCrunch test.
> > For my implementation, I created an array of servers in my Main class. I also created Event classes (e.g. ArriveEvent, ServeEvent), which has the method nextEvent that takes in a array of servers, modifies the server in the array and returns the next event. By calling .nextEvent(Server[] servers) in the main class, I am hence able to modify the array of servers that is in my Main class.
> 
> Hi! I did this, but I have a cyclic dependency between my Event class and Server class. This you experience something similar, and if so how did you fix it? Thanks!

for my implementation, only my event has a server. my server does not have an event. perhaps your server has an event & your event has a server?

**Comment by sebin0817** [comment]
> > From my experience, we are not allowed to have any attributes that are not private final, as it will fail the CodeCrunch test.
> > For my implementation, I created an array of servers in my Main class. I also created Event classes (e.g. ArriveEvent, ServeEvent), which has the method nextEvent that takes in a array of servers, modifies the server in the array and returns the next event. By calling .nextEvent(Server[] servers) in the main class, I am hence able to modify the array of servers that is in my Main class.
> 
> Hi! I did this, but I have a cyclic dependency between my Event class and Server class. This you experience something similar, and if so how did you fix it? Thanks!

You can try making a ServersList class that handles the array of servers! I took this in in my Main.java class and so far, i have not faced any cyclic dependencies yet. all the best!
